modules/cstrike/cstrike/CstrikeHacks.cpp

```diff
@@ -207,182 +219,199 @@ DETOUR_DECL_MEMBER0(GiveDefaultItems, void)  // void CBasePlayer::GiveDefaultIte
+	}
 
-	// Give me my item!
-	DETOUR_MEMBER_CALL(GiveNamedItem)(pszName);
+	return canBuy;
 }
 
-DETOUR_DECL_MEMBER1(GiveShield, void, bool, bRetire) // void CBasePlayer::GiveShield(bool bRetire)
+DETOUR_DECL_STATIC2(CanBuyThis, bool, void*, pvPlayer, int, weaponId) // bool CanBuyThis(CBasePlayer *pPlayer, int weaponId)
 {
-	// Special case for shield. Game doesn't use GiveNamedItem() to give a shield.
-	if (CurrentItemId == CSI_SHIELDGUN)
+	auto canBuy = DETOUR_STATIC_CALL(CanBuyThis)(pvPlayer, weaponId);
+
+	if (!canBuy || !TriggeredFromCommand || !((1 << CurrentItemId & CSI_ALL_GUNS) || CurrentItemId == CSI_SHIELD))
 	{
-		int client = TypeConversion.cbase_to_id(this);
+		return canBuy;
+	}
+
+	auto playerId = TypeConversion.cbase_to_id(pvPlayer);
 
-		if (MF_IsPlayerAlive(client) && MF_ExecuteForward(ForwardOnBuy, static_cast<cell>(client), CSI_SHIELDGUN) > 0)
+	if (MF_IsPlayerAlive(playerId) && get_pdata<int>(pvPlayer, MoneyDesc.fieldOffset) >= ItemsManager.GetItemPrice(CurrentItemId))
+	{
+		if (MF_ExecuteForward(ForwardOnBuy, playerId, CurrentItemId) > 0)
 		{
-			return;
+			canBuy = false;
 		}
 	}
 
-	// From here, forward is not blocked, resetting this
-	// to ignore code in AddAccount which is called right after.
-	CurrentItemId = 0;
+	return canBuy;
+}
+
+DETOUR_DECL_STATIC3(BuyGunAmmo, bool, void*, player, int, nSlot, bool, bBlinkMoney) // bool BuyGunAmmo(CBasePlayer *player, int nSlot, bool bBlinkMoney)
+{
+	auto result = DETOUR_STATIC_CALL(BuyGunAmmo)(player, nSlot, bBlinkMoney);
+
+	if (result && BlockAmmosUpdate)
+	{
+		BlockAmmosUpdate = false;
+		return false;
+	}
+
+	return result;
+}
+
+DETOUR_DECL_MEMBER1(GiveNamedItem, void, const char*, pszName) // void CBasePlayer::GiveNamedItem(const char *pszName)
+{
+	if (TriggeredFromCommand)
+	{
+		switch (CurrentItemId)
+		{
+			case CSI_VEST:
+			case CSI_VESTHELM:
+			case CSI_FLASHBANG:
+			case CSI_HEGRENADE:
+			case CSI_SMOKEGRENADE:
+			case CSI_PRIAMMO:
+			case CSI_SECAMMO:
+			{
+				auto playerId = TypeConversion.cbase_to_id(this);
 
-	// Give me my shield!
-	DETOUR_MEMBER_CALL(GiveShield)(bRetire);
+				if (MF_IsPlayerAlive(playerId) && MF_ExecuteForward(ForwardOnBuy, playerId, CurrentItemId) > 0)
+				{
+					BlockAmmosUpdate = CurrentItemId == CSI_PRIAMMO || CurrentItemId == CSI_SECAMMO;
+					BlockMoneyUpdate = true;
+					return;
+				}
+			}
+		}
+	}
+
+	DETOUR_MEMBER_CALL(GiveNamedItem)(pszName);
 }
 
 DETOUR_DECL_MEMBER2(AddAccount, void, int, amount, bool, bTrackChange) // void CBasePlayer::AddAccount(int amount, bool bTrackChange)
 {
-	// No buy command or forward not blocked.
-	// Resuming game flow.
-	if (!CurrentItemId)
+	if (BlockMoneyUpdate)
 	{
-		DETOUR_MEMBER_CALL(AddAccount)(amount, bTrackChange);
+		BlockMoneyUpdate = false;
+		return;
 	}
-	// Shield is blocked.
-	// We need to hook EmitSound to block pickup sound played right after.
-	else if (CurrentItemId == CSI_SHIELDGUN)
+
+	DETOUR_MEMBER_CALL(AddAccount)(amount, bTrackChange);
```

Comment:
Not sure why s1lent broke the compatibility with the original version.

Not something you can fix with gamedata alone, the module needs to be updated. That's not going to happen, for now, therefore your choices are either you add yourself the missing argument where the function is detoured in the code, you use ReGameDLL API, or as said above you're using ReAPI.